modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/indexreader/IgniteIndexReader.java

```diff
@@ -1219,6 +1241,44 @@ private TreeNode getTreeNode(long pageId, TreeTraverseContext nodeCtx) {
         }
     }
 
+    /**
+     * @param nodeCtx tree traverse context.
+     * @param pageContent inner page content.
+     * @param io inner page io.
+     * @throws IgniteCheckedException
+     */
+    private void checkTriangleInvariant(TreeTraverseContext nodeCtx, PageContent pageContent, PageIO io)
+        throws IgniteCheckedException {
+        if (io instanceof BPlusInnerIO && pageContent.linkedPageIds != null) {
+            long rightChildPageId = -1;
+
+            final ByteBuffer innerBuf = allocateBuffer(pageSize);
+
+            try {
+                // linkedPageIds are populated in InnerPageIOProcessor.getContent
+                for (Long linkedPageId : pageContent.linkedPageIds) {
+                    triangleChecksNum++;
+
+                    innerBuf.rewind();
+
+                    if (rightChildPageId != -1 && linkedPageId != rightChildPageId)
+                        triangleErrorsCounter++;
+
+                    readPage(nodeCtx.store, linkedPageId, innerBuf);
+
+                    final long childAddr = bufferAddress(innerBuf);
+
+                    final PageIO childIO = PageIO.getPageIO(childAddr);
+
+                    rightChildPageId = ((BPlusIO)childIO).getForward(childAddr);
+                }
+            }
+            finally {
+                freeBuffer(innerBuf);
```

Comment:
Is it possible to allocate a single buffer per `IgniteIndexReader`?

Reply:
I tried to do it. A single buffer is definitely not enough. After shooting my foot a couple of times I did some optimizations with class-wize buffer. But the usage of such buffer seems very error prune - you have to be always aware of who writes there, who rewinds, where the addresses of this buffer are still in use. So I propose to leave it as it is - use a separate buffer for every method that needs it. It will lead to several extra allocations but will protect from some non-trivial buffer misuses.